File: 12/gemini-3.0-pro/python/solution.py
import sys
import os

sys.setrecursionlimit(5000)

def solve():
    input_file = "input.txt"
    if len(sys.argv) > 1:
        input_file = sys.argv[1]

    if not os.path.exists(input_file):
        print(f"Input file {input_file} not found.")
        return

    with open(input_file) as f:
        content = f.read().replace("\r\n", "\n")

    parts = content.split("\n\n")
    shapes = {}
    regions = []

    for part in parts:
        lines = part.strip().split("\n")
        if not lines: continue

        if ":" in lines[0]:
            if "x" in lines[0]: # Region
                for line in lines:
                    if line.strip():
                        parse_region(line, regions)
            else: # Shape
                parse_shape(lines, shapes)

    solved_count = 0
    for region in regions:
        if solve_region(region, shapes):
            solved_count += 1

    print(f"Day 12 Part 1: {solved_count}")
    # print(f"Day 12 Part 2: 0")

def parse_shape(lines, shapes):
    header = lines[0].strip().rstrip(":")
    try:
        id = int(header)
    except ValueError:
        return

    points = set()
    for r in range(1, len(lines)):
        for c in range(len(lines[r])):
            if lines[r][c] == '#':
                points.add((r - 1, c))

    shapes[id] = generate_variations(points)

def parse_region(line, regions):
    p = line.split(":")
    dims = p[0].split("x")
    w = int(dims[0])
    h = int(dims[1])
    counts = [int(x) for x in p[1].strip().split()]
    regions.append({"w": w, "h": h, "counts": counts})

def generate_variations(original):
    variations = []
    current = original
    for i in range(2): # Flip
        for j in range(4): # Rotate
            normalized = normalize(current)
            if normalized not in variations:
                variations.append(normalized)
            current = rotate(current)
        current = flip(original)
    return variations

def rotate(points):
    return set((c, -r) for r, c in points)

def flip(points):
    return set((r, -c) for r, c in points)

def normalize(points):
    if not points: return frozenset()
    min_r = min(p[0] for p in points)
    min_c = min(p[1] for p in points)
    return frozenset((r - min_r, c - min_c) for r, c in points)

def solve_region(region, shapes):
    presents = []
    total_area = 0
    for i, count in enumerate(region["counts"]):
        for _ in range(count):
            presents.append(i)
            # Area of shape i is len(shapes[i][0])
            total_area += len(shapes[i][0])

    if total_area > region["w"] * region["h"]:
        return False

    max_skips = (region["w"] * region["h"]) - total_area
    grid = [False] * (region["w"] * region["h"])

    # Sort present types by size descending
    present_types = sorted(list(set(presents)), key=lambda x: len(shapes[x][0]), reverse=True)
    counts = {}
    for p in presents:
        counts[p] = counts.get(p, 0) + 1

    anchored_shapes = {}
    for type_id in present_types:
        anchored_shapes[type_id] = []
        for v in shapes[type_id]:
            # Sort points to find anchor
            sorted_pts = sorted(list(v))
            anchor = sorted_pts[0]
            points = [(p[0] - anchor[0], p[1] - anchor[1]) for p in sorted_pts]
            anchored_shapes[type_id].append(points)

    ctx = {
        "w": region["w"],
        "h": region["h"],
        "counts": counts,
        "types": present_types,
        "anchored_shapes": anchored_shapes
    }

    return backtrack(0, grid, ctx, max_skips)

def backtrack(idx, grid, ctx, skips_left):
    while idx < len(grid) and grid[idx]:
        idx += 1

    if idx == len(grid):
        return all(c == 0 for c in ctx["counts"].values())

    r = idx // ctx["w"]
    c = idx % ctx["w"]

    for type_id in ctx["types"]:
        if ctx["counts"][type_id] > 0:
            variations = ctx["anchored_shapes"][type_id]
            for shape in variations:
                fits = True
                for dr, dc in shape:
                    pr, pc = r + dr, c + dc
                    if pr < 0 or pr >= ctx["h"] or pc < 0 or pc >= ctx["w"]:
                        fits = False
                        break
                    if grid[pr * ctx["w"] + pc]:
                        fits = False
                        break

                if fits:
                    # Place
                    for dr, dc in shape:
                        grid[(r + dr) * ctx["w"] + (c + dc)] = True
                    ctx["counts"][type_id] -= 1

                    if backtrack(idx + 1, grid, ctx, skips_left):
                        return True

                    # Backtrack
                    ctx["counts"][type_id] += 1
                    for dr, dc in shape:
                        grid[(r + dr) * ctx["w"] + (c + dc)] = False

    if skips_left > 0:
        grid[idx] = True
        if backtrack(idx + 1, grid, ctx, skips_left - 1):
            return True
        grid[idx] = False

    return False

if __name__ == "__main__":
    solve()

File: 12/gemini-3.0-pro/typescript/solution.ts
import * as fs from 'fs';

function solve() {
    let inputFile = 'input.txt';
    if (process.argv.length > 2) {
        inputFile = process.argv[2];
    }

    if (!fs.existsSync(inputFile)) {
        console.log(`Input file ${inputFile} not found.`);
        return;
    }

    const content = fs.readFileSync(inputFile, 'utf-8').replace(/\r\n/g, '\n');
    const parts = content.split('\n\n');

    const shapes: { [id: number]: Set<string>[] } = {};
    const regions: RegionTask[] = [];

    for (const part of parts) {
        const lines = part.trim().split('\n');
        if (lines.length === 0) continue;

        if (lines[0].includes(':')) {
            if (lines[0].includes('x')) { // Region
                for (const line of lines) {
                    if (line.trim()) parseRegion(line, regions);
                }
            } else { // Shape
                parseShape(lines, shapes);
            }
        }
    }

    let solvedCount = 0;
    for (const region of regions) {
        if (solveRegion(region, shapes)) {
            solvedCount++;
        }
    }

    console.log(`Day 12 Part 1: ${solvedCount}`);
    // console.log(`Day 12 Part 2: 0`);
}

interface RegionTask {
    w: number;
    h: number;
    counts: number[];
}

function parseShape(lines: string[], shapes: { [id: number]: Set<string>[] }) {
    const header = lines[0].trim().replace(':', '');
    const id = parseInt(header);
    if (isNaN(id)) return;

    const points = new Set<string>();
    for (let r = 1; r < lines.length; r++) {
        for (let c = 0; c < lines[r].length; c++) {
            if (lines[r][c] === '#') {
                points.add(`${r - 1},${c}`);
            }
        }
    }
    shapes[id] = generateVariations(points);
}

function parseRegion(line: string, regions: RegionTask[]) {
    const p = line.split(':');
    const dims = p[0].split('x');
    const w = parseInt(dims[0]);
    const h = parseInt(dims[1]);
    const counts = p[1].trim().split(/\s+/).map(x => parseInt(x));
    regions.push({ w, h, counts });
}

function generateVariations(original: Set<string>): Set<string>[] {
    const variations: Set<string>[] = [];
    let current = original;
    for (let i = 0; i < 2; i++) { // Flip
        for (let j = 0; j < 4; j++) { // Rotate
            const normalized = normalize(current);
            if (!variations.some(v => setEquals(v, normalized))) {
                variations.push(normalized);
            }
            current = rotate(current);
        }
        current = flip(original);
    }
    return variations;
}

function rotate(points: Set<string>): Set<string> {
    const newPoints = new Set<string>();
    for (const p of points) {
        const [r, c] = p.split(',').map(Number);
        newPoints.add(`${c},${-r}`);
    }
    return newPoints;
}

function flip(points: Set<string>): Set<string> {
    const newPoints = new Set<string>();
    for (const p of points) {
        const [r, c] = p.split(',').map(Number);
        newPoints.add(`${r},${-c}`);
    }
    return newPoints;
}

function normalize(points: Set<string>): Set<string> {
    if (points.size === 0) return points;
    let minR = Infinity;
    let minC = Infinity;
    for (const p of points) {
        const [r, c] = p.split(',').map(Number);
        if (r < minR) minR = r;
        if (c < minC) minC = c;
    }
    const newPoints = new Set<string>();
    for (const p of points) {
        const [r, c] = p.split(',').map(Number);
        newPoints.add(`${r - minR},${c - minC}`);
    }
    return newPoints;
}

function setEquals(a: Set<string>, b: Set<string>): boolean {
    if (a.size !== b.size) return false;
    for (const p of a) {
        if (!b.has(p)) return false;
    }
    return true;
}

interface AnchoredShape {
    points: { dr: number, dc: number }[];
}

interface SolverContext {
    w: number;
    h: number;
    counts: { [id: number]: number };
    types: number[];
    anchoredShapes: { [id: number]: AnchoredShape[] };
}

function solveRegion(region: RegionTask, shapes: { [id: number]: Set<string>[] }): boolean {
    const presents: number[] = [];
    let totalArea = 0;
    for (let i = 0; i < region.counts.length; i++) {
        for (let j = 0; j < region.counts[i]; j++) {
            presents.push(i);
            totalArea += shapes[i][0].size;
        }
    }

    if (totalArea > region.w * region.h) return false;

    const maxSkips = (region.w * region.h) - totalArea;
    const grid = new Uint8Array(region.w * region.h); // 0: empty, 1: filled

    const presentTypes = Array.from(new Set(presents)).sort((a, b) => shapes[b][0].size - shapes[a][0].size);
    const counts: { [id: number]: number } = {};
    for (const p of presents) {
        counts[p] = (counts[p] || 0) + 1;
    }

    const anchoredShapes: { [id: number]: AnchoredShape[] } = {};
    for (const type of presentTypes) {
        anchoredShapes[type] = [];
        for (const v of shapes[type]) {
            const sortedPts = Array.from(v).map(p => {
                const [r, c] = p.split(',').map(Number);
                return { r, c };
            }).sort((a, b) => a.r - b.r || a.c - b.c);

            const anchor = sortedPts[0];
            const points = sortedPts.map(p => ({ dr: p.r - anchor.r, dc: p.c - anchor.c }));
            anchoredShapes[type].push({ points });
        }
    }

    const ctx: SolverContext = {
        w: region.w,
        h: region.h,
        counts,
        types: presentTypes,
        anchoredShapes
    };

    return backtrack(0, grid, ctx, maxSkips);
}

function backtrack(idx: number, grid: Uint8Array, ctx: SolverContext, skipsLeft: number): boolean {
    while (idx < grid.length && grid[idx] !== 0) {
        idx++;
    }

    if (idx === grid.length) {
        return Object.values(ctx.counts).every(c => c === 0);
    }

    const r = Math.floor(idx / ctx.w);
    const c = idx % ctx.w;

    for (const type of ctx.types) {
        if (ctx.counts[type] > 0) {
            const variations = ctx.anchoredShapes[type];
            for (const shape of variations) {
                let fits = true;
                for (const pt of shape.points) {
                    const pr = r + pt.dr;
                    const pc = c + pt.dc;
                    if (pr < 0 || pr >= ctx.h || pc < 0 || pc >= ctx.w) { fits = false; break; }
                    if (grid[pr * ctx.w + pc] !== 0) { fits = false; break; }
                }

                if (fits) {
                    for (const pt of shape.points) {
                        grid[(r + pt.dr) * ctx.w + (c + pt.dc)] = 1;
                    }
                    ctx.counts[type]--;

                    if (backtrack(idx + 1, grid, ctx, skipsLeft)) return true;

                    ctx.counts[type]++;
                    for (const pt of shape.points) {
                        grid[(r + pt.dr) * ctx.w + (c + pt.dc)] = 0;
                    }
                }
            }
        }
    }

    if (skipsLeft > 0) {
        grid[idx] = 1;
        if (backtrack(idx + 1, grid, ctx, skipsLeft - 1)) return true;
        grid[idx] = 0;
    }

    return false;
}

solve();
